### PR TITLE
Develop pre

### PR DESCRIPTION
**Pull request by rajasekari2i**

### CODE REVIEW

File: controller/medication_controller.py [hunk 1]
```diff
@@ -61,19 +61,35 @@ def create_medication_request(med: MedicationCreateModel, patient_id: str):
...
+            request = response_request.json()
+            result["is_other_medication_exist"] = True if request.get("total", 0) > 0 else False
+
+            if med.request_approved and not result["is_other_medication_exist"]:
                 medication_request = MedicationClient.create_medication_request(
                     med, patient_id
                 )
                 medication_request.save()
                 result["other_medication_id"] = medication_request.id
+            result["is_other_medication_exist"] = True
```
Comment (tamsud) [inline controller/medication_controller.py:84]: will it be true always or when it become  false

File: controller/medication_controller.py [hunk 2]
```diff
@@ -61,19 +61,35 @@ def create_medication_request(med: MedicationCreateModel, patient_id: str):
...
 
-            if med.statement_approved:
+            if med.statement_approved and not result["is_current_medication_exist"]:
                 medication_statement = MedicationClient.create_medication_statement(
                     med, patient_id
                 )
                 medication_statement.save()
                 result["current_medication_id"] = medication_statement.id
+            result["is_current_medication_exist"] = True
```
Comment (tamsud) [inline controller/medication_controller.py:92]: will it be true always or when it become false

File: controller/patient_controller.py [hunk 3]
```diff
@@ -33,46 +39,65 @@ class PatientClient:
...
             )
 
             if Patient.get({"id": patient_id_update}):
                 return HTTPException(
                     detail=f"Error: Patient with id {patient_id_update} already exists",
                     status_code=status.HTTP_400_BAD_REQUEST,
                 )
+            result = {}
+
+            # if not pat.createAccount:
```
Comment (tamsud) [inline controller/patient_controller.py:60]: why code is commented out here

File: controller/patient_controller.py [hunk 4]
```diff
@@ -159,7 +207,15 @@ def get_all_patients(page, page_size):
         try:
             patients = paginate(Patient, page, page_size)
             logger.info(f"Patients Found: {len(patients)}")
-            return patients
+            if patients.get('total', 1) == 0:
```
Comment (tamsud) [inline controller/patient_controller.py:210]: why default is 1

File: controller/service_history_controller.py [hunk 5]
```diff
@@ -11,22 +11,36 @@
...
             if immunization:
-                return ServiceHistoryClient.process_immunization(immuzation)
+                immuzation = HL7ImmunizationClient.get_immunizations_by_patient_id(patient_id, page, count)
+
+            if all_service:
+                service_history = ObservationClient.get_lab_result_by_patient_id(patient_id, page, count)
+                immuzation = HL7ImmunizationClient.get_immunizations_by_patient_id(patient_id, page, count)
+                return ServiceHistoryClient.create_final_values(service_history, immuzation)
+            
+            if lab_result and not immunization:
```
Comment (tamsud) [inline controller/service_history_controller.py:26]: can you explain:

            if lab_result and not immunization:
                if not service_history:  
                    return []
                else:
                    return ServiceHistoryClient.process_lab_result(service_history)

            if immunization and not lab_result:
                if not immuzation:  
                    return []
                else:
                    return ServiceHistoryClient.process_immunization(immuzation)

File: models/encounter_validation.py [hunk 6]
```diff
@@ -2,9 +2,9 @@
 from pydantic import BaseModel, validator
 
 def validate_phone_number(phone_number: str) -> str:
-    if not re.match(r'^\+1\d{10}$', phone_number):
-        raise ValueError('Phone number must be in the format +1 followed by 10 digits')
-    return phone_number
+    if not re.match(r'^\d{10}$', phone_number):
+        raise ValueError('Phone number must be 10 digits')
+    return f"+1{phone_number}"
```
Comment (tamsud) [inline models/encounter_validation.py:7]: why are we hardcoding +1 while return the ph number

File: models/patient_validation.py [hunk 7]
```diff
@@ -2,87 +2,111 @@
...
-    except ValueError:
-        raise ValueError('date_of_birth must be in the format YYYY-MM-DD')
-    return date_str
+        utc_dt = datetime.fromtimestamp(timestamp, tz=timezone.utc)
+    except (OverflowError, ValueError) as e:
+        return {"Error": str(e)}
+    return utc_dt.strftime('%Y-%m-%d')
 
 
 def validate_phone_number(phone_number: str) -> str:
```
Comment (tamsud) [inline models/patient_validation.py:22]: models/encounter_validation.py same fucntion there in two places

remove one of them

File: routes/appointment_routes.py [hunk 8]
```diff
@@ -46,16 +44,23 @@ async def get_by_id(
...
-    logger.info(f"fetch appointment for {appointment_by_name}")
-    return AppointmentClient.get_by_patient_name(appointment_by_name, page, page_size)
+async def get_appointment(
+        request: Request,
+        patient_id: Optional[str] = None,
+        name:  Optional[str] = None,
+        appointment_by_name: Optional[str] = None,
+        state_date: Optional[str] = None,
+        end_date: Optional[str] = None,
+        all_appointment: Optional[bool] = None,
```
Comment (tamsud) [inline routes/appointment_routes.py:56]: why we need all_appointment boolean? by default we can go with all appointment right

**Review by tamsud — CHANGES_REQUESTED** [review]

File: controller/appointment_controller.py [hunk 9]
```diff
@@ -33,9 +46,19 @@ def create_appointment(patient_id: str, app: AppoinmentModel):
                         status=PARTICIPANT_STATUS,
                     )
                 ],
-                start=app.date_of_appoinment,
+                start=app.date_of_appointment,
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/appointment_controller.py:93]: changed it

File: controller/encounter_controller.py [hunk 10]
```diff
@@ -213,14 +215,18 @@ def update_by_patient_id(patient_id: str, encounter_id: str, enc: EncounterUpdat
...
     @staticmethod
-    def get_all_encounters():
+    def get_all_encounters(page, page_size):
         try:
-            encounters = Encounter.get()
-            if encounters:
-                logger.info(f"Encounters Found {len(encounters)}")
-                return encounters
+            encounters = paginate(Encounter, page, page_size)
+            if encounters.get('total', 1) == 0:
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/encounter_controller.py:258]: changed the default

File: controller/insurance_controller.py [hunk 11]
```diff
@@ -6,18 +6,65 @@
...
 logger = logging.getLogger("log")
 
 
 class CoverageClient:
     @staticmethod
-    def create_coverage(coverage: CoverageModel, patient_id: str):
+    def create_coverage(coverage: PatientModel, patient_id: str, from_patient=False):
+        try:
+            result = {}
+            primary_insurance = CoverageWrapper if from_patient else Coverage
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py:34]: now changed that to single variable and used it

File: controller/insurance_controller.py [hunk 12]
```diff
@@ -26,25 +73,33 @@ def create_coverage(coverage: CoverageModel, patient_id: str):
...
-            )
-            insurance_plan.save()
-            response_data = {"id": insurance_plan.id, "created": True}
-            logger.info(f"Added Successfully in DB: {response_data}")
-            return response_data
+            if existing_coverages:
+                primary_coverage_result = CoverageClient.get_primary_coverage(existing_coverages, patient_id)
+                secondary_coverage_result = CoverageClient.get_secondary_coverage(existing_coverages, patient_id)
+                tertiary_coverage_result = CoverageClient.get_tertiary_coverage(existing_coverages, patient_id)
+                if primary_coverage_result.get('is_primary_insurance_exist') is True:
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py:81]: changed the condition

File: controller/insurance_controller.py [hunk 13]
```diff
@@ -26,25 +73,33 @@ def create_coverage(coverage: CoverageModel, patient_id: str):
...
+                    result["is_primary_coverage_exist"] = True
+                if primary_coverage_result.get('is_primary_insurance_exist') is False and ins_plan.insurance_type == "primary":
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py]: changed the condition

File: controller/insurance_controller.py [hunk 14]
```diff
@@ -26,25 +73,33 @@ def create_coverage(coverage: CoverageModel, patient_id: str):
...
+                    primary_insurance_plan = CoverageClient.create_insurance_plan(patient_id, ins_plan)
+                    primary_insurance_plan.save()
+                    result["is_primary_insurance"] = primary_insurance_plan.id
+                if secondary_coverage_result.get('is_secondary_insurance_exist') is True:
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py:88]: changed the condition

File: controller/insurance_controller.py [hunk 15]
```diff
@@ -26,25 +73,33 @@ def create_coverage(coverage: CoverageModel, patient_id: str):
...
+                    result["is_secondary_coverage_exist"] = True
+                if secondary_coverage_result.get('is_secondary_insurance_exist') is False and ins_plan.insurance_type == "secondary":
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py]: changed the condition

File: controller/insurance_controller.py [hunk 16]
```diff
@@ -26,25 +73,33 @@ def create_coverage(coverage: CoverageModel, patient_id: str):
...
+                    secondary_insurance_plan = CoverageClient.create_insurance_plan(patient_id, ins_plan)
+                    secondary_insurance_plan.save()
+                    result["is_secondary_insurance"] = secondary_insurance_plan.id
+                if tertiary_coverage_result.get("is_tertiary_insurance_exist") is True:
+                    result["is_tertiary_insurance_exist"] = True
+                if tertiary_coverage_result.get('is_tertiary_insurance_exist') is False and ins_plan.insurance_type == "tertiary":
+                    tertiary_insurance_plan = CoverageClient.create_insurance_plan(patient_id, ins_plan)
+                    tertiary_insurance_plan.save()
+                    result["is_tertiary_insurance"] = tertiary_insurance_plan.id
+            result["created"] = True
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py:102]: if its created mean we will return created = true. i kept it as common

File: controller/insurance_controller.py [hunk 17]
```diff
@@ -111,41 +164,72 @@ def get_coverage_by_id(patient_id: str, insurance_id: str):
...
+            if patient_id and insurance_id:
+                response_coverage = Coverage.make_request(method="GET", endpoint=f"/fhir/Coverage/{insurance_id}?beneficiary=Patient/{patient_id}")
+                coverage = response_coverage.json() 
+                patient_id_occurrences = sum(1 for entry in coverage.get('entry', []) if entry['resource']['beneficiary']['reference'] == f"Patient/{patient_id}")
+                if patient_id_occurrences >= 3:
+                    logger.error(f"A patient can only have 3 insurance")
+                    return JSONResponse(
+                        content="A patient can only have 3 insurance", status_code=status.HTTP_200_OK
+                    )
+                if response_coverage.status_code == 404:
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py:205]: changed, firstly checking the validation and put this logic

File: controller/insurance_controller.py [hunk 18]
```diff
@@ -111,41 +164,72 @@ def get_coverage_by_id(patient_id: str, insurance_id: str):
...
+            else:
+                response_coverage = Coverage.make_request(method="GET", endpoint=f"/fhir/Coverage/?beneficiary=Patient/{patient_id}")
+                coverages = response_coverage.json()
+                patient_id_occurrences = sum(1 for entry in coverages.get('entry', []) if entry['resource']['beneficiary']['reference'] == f"Patient/{patient_id}")
+                if patient_id_occurrences >= 3:
+                    logger.error(f"A patient can only have 3 insurance")
+                    return JSONResponse(
+                        content="A patient can only have 3 insurance", status_code=status.HTTP_400_BAD_REQUEST
+                    )
+                if response_coverage.status_code == 404:
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py:225]: changed, firstly checking the validation and put this logic

File: controller/insurance_controller.py [hunk 19]
```diff
@@ -111,41 +164,72 @@ def get_coverage_by_id(patient_id: str, insurance_id: str):
...
+                    logger.info(f"Coverage Not Found: {patient_id}")
+                    return JSONResponse(
+                        content={"error": "No Matching Record"},
+                        status_code=status.HTTP_404_NOT_FOUND
+                    )
+                if coverages:
+                    primary_coverage_result = CoverageClient.get_primary_coverage(coverages, patient_id)
+                    secondary_coverage_result = CoverageClient.get_secondary_coverage(coverages, patient_id)
+                    tertiary_coverage_result = CoverageClient.get_tertiary_coverage(coverages, patient_id)
+                if primary_coverage_result.get('is_primary_insurance_exist') is True:
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py:236]: changed the condition

File: controller/insurance_controller.py [hunk 20]
```diff
@@ -111,41 +164,72 @@ def get_coverage_by_id(patient_id: str, insurance_id: str):
...
+                    result["is_primary_coverage_exist"] = True
+                if primary_coverage_result.get('is_primary_insurance_exist') is False and updated_coverage.insurance_type== "primary":
+                    primary_insurance_plan = CoverageClient.create_insurance_plan(patient_id, updated_coverage)
+                    primary_insurance_plan.save()
+                    result["primary_insurance_id"] = primary_insurance_plan.id
+                if secondary_coverage_result.get('is_secondary_insurance_exist') is True:
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py:242]: changed the condition

File: controller/insurance_controller.py [hunk 21]
```diff
@@ -111,41 +164,72 @@ def get_coverage_by_id(patient_id: str, insurance_id: str):
...
+                    result["is_secondary_coverage_exist"] = True
+                if secondary_coverage_result.get('is_secondary_insurance_exist') is False and updated_coverage.insurance_type== "secondary":
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py:244]: changed the condition

File: controller/insurance_controller.py [hunk 22]
```diff
@@ -111,41 +164,72 @@ def get_coverage_by_id(patient_id: str, insurance_id: str):
...
+                    secondary_insurance_plan = CoverageClient.create_insurance_plan(patient_id, updated_coverage)
+                    secondary_insurance_plan.save()
+                    result["secondary_insurance_id"] = secondary_insurance_plan.id
+                if tertiary_coverage_result.get("is_tertiary_insurance_exist") is True:
+                    result["is_tertiary_insurance_exist"] = True
+                if tertiary_coverage_result.get('is_tertiary_insurance_exist') is False and updated_coverage.insurance_type== "tertiary":
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py:250]: changed the condition

File: controller/insurance_controller.py [hunk 23]
```diff
@@ -197,3 +281,129 @@ def delete_by_insurance_id(insurance_id: str, patient_id: str):
                 status_code=status.HTTP_400_BAD_REQUEST
             )
 
+    @staticmethod
+    def get_primary_coverage(existing_coverages, patient_id):
+        result = {}
+        for entry in existing_coverages.get("entry", []):
+            resource = entry["resource"]
+            beneficiary_id = resource["beneficiary"]["reference"].split("/")[-1]
+            if beneficiary_id == patient_id:
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py:329]: just checking both the id is same

File: controller/insurance_controller.py [hunk 24]
```diff
@@ -197,3 +281,129 @@ def delete_by_insurance_id(insurance_id: str, patient_id: str):
...
+                for coverage_class in resource.get("class", []):
+                    if coverage_class.get("value") == "Tertiary":
+                        result["is_tertiary_insurance_exist"] = True
+                        return result
+        result["is_tertiary_insurance_exist"] = False
+        return result
+
+    @staticmethod
+    def create_primary_insurance(primary_insurance, coverage: PatientModel, patient_id: str):
+        insurance_status = "active" if coverage.haveInsurance else "draft"
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py:367]: removed one field and kept only the status

File: controller/insurance_controller.py [hunk 25]
```diff
@@ -197,3 +281,129 @@ def delete_by_insurance_id(insurance_id: str, patient_id: str):
...
+                    type=CodeableConcept(coding=[Coding(system=GROUP_SYSTEM, code=GROUP_CODE)]),
+                    value=updated_coverage.insurance_type,
+                    name=updated_coverage.groupNumber
+                )
+            ]
+        )
+    
+    @staticmethod
+    def create_insurance_plan(patient_id, updated_coverage: CoverageModel):
+        result = PRIMARY_STATUS if updated_coverage.insurance_type == "primary" else SECONDARY_STATUS \
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/insurance_controller.py:431]: changed and checked with lower()

File: controller/lab_result_controller.py [hunk 26]
```diff
@@ -35,12 +36,19 @@ def get_lab_result_by_id(patient_id: str, lab_result_id: str):
             )
         
     @staticmethod
-    def get_all_lab_result():
+    def get_all_lab_result(page, page_size):
         try:
-            lab_result = Observation.get()
-            if lab_result:
+            lab_result = paginate(Observation, page, page_size)
+            if lab_result.get('total', 1) == 0:
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/lab_result_controller.py:50]: changed the default value

File: controller/master_controller.py [hunk 27]
```diff
@@ -8,7 +8,7 @@ class MasterClient:
     @staticmethod
     def get_master_data(table_name: str):
         table = {"race": "CimparRace", "state": "CimparState", "lab_test": "CimparLabTest",
-                 "ethnicity": "CimparEthnicity"}
+                 "ethnicity": "CimparEthnicity", "company" : "CimparInsuranceCompany"}
```
[the remark this answers is not in the document]
Comment (MonishaSivanathan) [reply controller/master_controller.py]: updated

File: controller/medication_controller.py (hunk 2 above)
Comment (MonishaSivanathan) [reply controller/medication_controller.py:92]: remoed the condition

File: controller/patient_controller.py (hunk 3 above)
Comment (MonishaSivanathan) [reply controller/patient_controller.py]: removed the code

File: controller/patient_controller.py (hunk 4 above)
Comment (MonishaSivanathan) [reply controller/patient_controller.py:226]: changed the default value

File: controller/service_history_controller.py (hunk 5 above)
Comment (MonishaSivanathan) [reply controller/service_history_controller.py]: we have removed the code and moved to aidbox query

File: models/encounter_validation.py (hunk 6 above)
Comment (MonishaSivanathan) [reply models/encounter_validation.py]: removed the hard coded value

File: models/patient_validation.py (hunk 7 above)
Comment (MonishaSivanathan) [reply models/patient_validation.py:31]: changed the function and have imported from patient_validation

**Review by tamsud — APPROVED** [review]

**Pull request merged**